### ISSUE TITLE
Error: Cannot find module 'package.json'

### ISSUE THREAD

**Issue by JounQin** [antonk52/lilconfig#17]
![image](https://user-images.githubusercontent.com/8336744/118075438-93e26280-b3e2-11eb-88f8-28502aae5a04.png)

It runs failed on GitLab runner.

![image](https://user-images.githubusercontent.com/8336744/118075501-a9578c80-b3e2-11eb-8947-4678676d78c8.png)

related https://github.com/shashkovdanil/clean-publish/issues/110

**Comment by JounQin** [comment]
![image](https://user-images.githubusercontent.com/8336744/118075551-ce4bff80-b3e2-11eb-8e8d-ceee62758cb8.png)

I can reproduce it on Node REPL.

**Comment by ai** [comment]
@antonk52 we have this issue in `clean-publish`

**Comment by antonk52** [comment]
Thank you for the issue! I will be able to take a closer look later on tonight, in the meanwhile can you add reproducible steps to the issue description? @JounQin

**Comment by JounQin** [comment]
I want to reproduce it too, but it only occurs on our GitLab CI runner sadly.

We're using `node:lts-alpine`.

It occurred when there is no config for `clean-publish`.

**Comment by antonk52** [comment]
From a brief look from my phone it looks like an issue with cwd when running the clean publish. You should get the same result if you’d use cosmiconfig since they work identical in this regard.

the comment with nodejs repl is not really useful here since it should be run in the same dir as the clean publish being executed in order to pick up the config or find it in package.json file. 

Once again sorry for a delayed reaction. Been off for a vacation. 

I’m going to have a closer look after work. Would you mind creating a repo where the issue is reproducible or linking to your project if it’s openly available?

**Comment by JounQin** [comment]
> From a brief look from my phone it looks like an issue with cwd when running the clean publish. You should get the same result if you’d use cosmiconfig since they work identical in this regard.

We're using other packages depends on `cosmiconfig`, but it never happens before.

> Would you mind creating a repo where the issue is reproducible or linking to your project if it’s openly available?

I will try to reproduce later, but it occurred on our private own GitLab service, so I have to figure out how to reproduce first. 😂

**Comment by antonk52** [comment]
@JounQin looking at how `clean-publish` [uses](https://github.com/shashkovdanil/clean-publish/blob/40b148f1f19f7596fe5839ca0fb5f2a5d83a9a3b/get-config.js#L83-L85) lilconfig

It would walk down the directories relative to where the command was executed, ie if you current working directory is `/Users/Foo/path/to/project`, it will try these paths in this order until it finds a config or returns null

<details><summary>paths:</summary>

- /Users/Foo/path/to/project/package.json

- /Users/Foo/path/to/project/.clean-publish
- /Users/Foo/path/to/project/.clean-publish.js
- /Users/Foo/path/to/package.json
- /Users/Foo/path/to/.clean-publish
- /Users/Foo/path/to/.clean-publish.js
- /Users/Foo/path/package.json
- /Users/Foo/path/.clean-publish
- /Users/Foo/path/.clean-publish.js
- /Users/Foo/package.json
- /Users/Foo/.clean-publish
- /Users/Foo/.clean-publish.js
</details >

From what I can see in the description the [check](https://github.com/antonk52/lilconfig/blob/4544c651f152fe96182650a78c8b763077265594/src/index.ts#L196-L200) for accessible file has passed and we try to require the file [here](https://github.com/antonk52/lilconfig/blob/4544c651f152fe96182650a78c8b763077265594/src/index.ts#L206).

The one thing that stands out for me is that the path in the logs provided is `Error: cannot find module "package.json"`. If we print [`searchItems`](https://github.com/antonk52/lilconfig/blob/4544c651f152fe96182650a78c8b763077265594/src/index.ts#L194) from the source code the paths are absolute(example `/Users/Foo/path/package.json`) instead of filenames only. 

This is as far as I can say from all the information provided so far. For further debugging and a potential fix I would need a reproducible example and more information about the environment where this error occurs, ie os and node version.

**Comment by JounQin** [comment]
It would be better to have a `debug` mode.

I'll try to use [`patch-package`](https://github.com/ds300/patch-package) for debugging first.

**Comment by antonk52** [comment]
as an alternative I could ask you to create a task in your CI (to run it once and share the log) with the code below. I've copied the code from how clean-publish uses lilconfig to see if it will throw an error if we call lilconfig explicitly. It also prints info about the current working directory and nodes version and OS that the script is being run on. This should provide enough info to narrow down what causes the error.

```js
const {lilconfig} = require('lilconfig')

console.log('common info', {
  cwd: process.cwd(),
  version: process.version,
  arch: process.arch,
  platform: process.platform,  
})

async main() {
  const result = await lilconfig('clean-publish', {searchPlaces: ['package.json', '.clean-publish', '.clean-publish.js']})
    .search()
    .catch(error => {
      console.log('search error:', error)
    })
  
  console.log({result})
  
  return result;  
}

main().catch(error => {
  console.log('main error:', error)
  process.exit(1)
})
```

**Comment by JounQin** [comment]
@antonk52 Sorry I was busy these days.

I've just tried your debug codes, and here is the logs:

```log
common info {
  cwd: '/builds/frontend/custom-webpack',
  version: 'v12.18.4',
  arch: 'x64',
  platform: 'linux'
}
search error: Error: Cannot find module 'package.json'
Require stack:
- /builds/frontend/custom-webpack/node_modules/lilconfig/dist/index.js
- /builds/frontend/custom-webpack/test.js
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:965:15)
    at Function.Module._load (internal/modules/cjs/loader.js:841:27)
    at Module.require (internal/modules/cjs/loader.js:1025:19)
    at require (internal/modules/cjs/helpers.js:72:18)
    at Object.search (/builds/frontend/custom-webpack/node_modules/lilconfig/dist/index.js:108:33)
    at async main (/builds/frontend/custom-webpack/test.js:11:18) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/builds/frontend/custom-webpack/node_modules/lilconfig/dist/index.js',
    '/builds/frontend/custom-webpack/test.js'
  ]
}
{ result: undefined }
```

**Comment by JounQin** [comment]
After using `patch-package`, I got `searchPaths` like the following:

```log
searchPaths: [
  '/builds/frontend/custom-webpack',
  '/builds/frontend',
  '/builds',
  ''
]
```

**Comment by jaydenseric** [comment]
Just a shot in the dark, but it looks like `lilconfig` uses `require` to load `package.json`:

https://github.com/antonk52/lilconfig/blob/4544c651f152fe96182650a78c8b763077265594/src/index.ts#L71
https://github.com/antonk52/lilconfig/blob/4544c651f152fe96182650a78c8b763077265594/src/index.ts#L205-L206

This is a risky thing to do these days, because sometimes people add a package [`exports`](https://nodejs.org/api/packages.html#packages_exports) field without allowing access to `package.json` (with and without the `.json` file extension) like this:

https://github.com/jaydenseric/graphql-upload/blob/bdb5f808e0d514c28c0f59c1abd71680aba29bae/package.json#L41-L42

Maybe this sort of problem relates to this issue? Regardless, it's a good idea to stop assuming that because a `.js` or `.json` file exists in a package that it can be read using `require`. Either:

- Prevent an error in this situation in `lilconfig`, and the burden lies on package authors to publish correct packages.
- Don't respect package `exports`, and use `fs.readFile` instead of `require` or `import` to access the contents of config files within packages.

Note that when this `lilconfig` migrates to pure ESM you won't be able to `import` `.json` files, and will be forced to use `fs.readFile` anyway.

**Comment by JounQin** [comment]
@jaydenseric No, the key problem here is `searchPaths` with resolve `''` if the cwd is not under `os.homedir()`, but it should be `'/'`, because `/package.json` is an absolute path, but `package.json` is not.

**Comment by antonk52** [comment]
@jaydenseric I've thought about loaders using require and don't think that's the case here. @JounQin already provided a fix in #19 which will be merged shortly. But I'd like to elaborate on the loaders for transparency.

Lilconfig tries to mimic cosmiconfig as much as possible. You can see the test cases to how it was tested to be as compatible with cosmiconfig at its best.

Cosmiconfig essentially uses the similar approaches to load js and json files
- for json it uses `parse-json` module in order to provide a more human friendly error messages, though it would throw for an invalid json the same as `require` does but with a different message. Please see this [test case](https://github.com/antonk52/lilconfig/blob/4544c651f152fe96182650a78c8b763077265594/src/spec/index.spec.ts#L599-L613) for an example
- for js files cosmiconfig uses `import-fresh` module to be able to require cached files as fresh ones. Internally it [uses](https://github.com/sindresorhus/import-fresh/blob/f1dddc06da4f1cdfb44a3aaba1f2c353c00dea02/index.js#L32) `require` also.

I would like to preserve the tight compatibility with cosmiconfig and avoid adding functionality that can diverge with cosmiconfig.

**Comment by antonk52** [comment]
This should be fixed in the [v2.0.3](https://github.com/antonk52/lilconfig/releases/tag/v2.0.3) release.

Feel free to reopen if the error occurs again.
